src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,548 @@
+//! Track corrupt/missing grid blocks.
+//!
+//! - The GridRepairQueue is LSM-aware: it can repair entire tables.
+//! - The GridRepairQueue is shared by all Trees.
+//! - The GridRepairQueue is "coherent" – that is, all of the blocks in the queue belong in the
+//!   replica's current checkpoint:
```

Comment:
:thinking: I think this is covered when grid writes complete: `assert(!grid.superblock.free_set.is_free(completed_write.address));`.

(But I'm also not really sure how this assertion you're suggesting would work, so possibly I'm missing something. Maybe it is something that could go in the `StorageChecker` (or a similar checker) instead?)

Reply:
What I am thinking here is that we use two "coordinates" to refer to blocks: address & checksum. But actually, there are three: address, checksum, and checkpoint_id.

We have an invariant that at most one checkpoint is active at any given point in time, but, as far as I can tell, this invariant is implicit. That is, when someone calls

```
pub fn enqueue_block(queue: *GridRepairQueue, address: u64, checksum: u128) void {
```

they call it at a particular `checkpoint_id`, and that `checkpoint_id` must match the current `checkpoint_id` of the `GridRepairQueue`. 

Similarly, calling `checkpoint_commence` implicitly advances this current `checkpoint_id`. 

The problem I see here is that this is very implicit --- we don't _actually_ pass the new `checkpoint_id`. If someone calls `enqueue_block` with arguments from a wrong checkpoint, this will probably trip some assert somewhere eventually, but that would be a generic "state of the grid is bad" assert we get on a subsequent read, not "you are trying to write data for checkpoint X, but the current checkpoint is Y" on the immediate problematic write. 

We do have `is_free` checks which I think solve part of the problem here. If I understand this correctly, a block at particular address evolves as `free -> checksumA -> free -> checksumB`, it can't evolve as `free -> checksumA -> -> checksumB`. So, it's impossible to confuse two neighbouring checkpoints, because they pass through a "free" state. But I think this ABA problem exists across several checkpoints, where a block has different contents at checkpoints `X` and `X+2`, but it is not free in either (but free in `X+1`). 

So, my overall thought here is to add

```zig
fn assert_coheret(checkpoint_id: u128, block_addr: u64, block_checksum: u128)
```

to record the fact that `block_addr` at `checkpoint_id` has contents `block_checksum`, and call that function whenever convenient, on write and read paths. Implementation wise, I think this assert could maintain a global backing hashmap keyed by `(addr, checkpoint)` tuple, which would `get_or_put` checksum and verify that it's never overwritten with conflicting values. 

---


To maybe arrive at the same point from a different angle, imagine writing an Ada-Spark style proof here that the code works correctly. I think that, to reason about correctness, we would actually have to both prove and also rely on the fact that all the faulty blocks in repair queue are coherent (belong to the same checkpoint). And to write such a proof, we'll probably need to introduce a "shadow state" --- a logical variable that tracks the current checkpoint, which is updated in the proof, but doesn't actually exist in code

And my thinking is that maybe we should actually go and materialize this shadow variable into an actual variable which tracks the current checkpoint from repair queue point of view?  Strictly speaking, this is redundant, but I think it might create space for more thorough assertions.